File: app/src/main/java/com/github/premnirmal/ticker/ui/StockFieldView.java
package com.github.premnirmal.ticker.ui;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.github.premnirmal.tickerwidget.R;

/**
 * Created by premnirmal on 3/4/15.
 */
public class StockFieldView extends LinearLayout {

    private TextView fieldName;
    private TextView fieldValue;

    public StockFieldView(Context context) {
        this(context, null);
    }

    public StockFieldView(Context context, AttributeSet attrs) {
        super(context, attrs);
        LayoutInflater.from(context).inflate(R.layout.stock_field_view, this, true);
        fieldName = ((TextView) findViewById(R.id.fieldname));
        fieldValue = ((TextView) findViewById(R.id.fieldvalue));
        if (attrs != null) {
            final TypedArray array = getContext().obtainStyledAttributes(attrs, R.styleable.StockFieldView);
            final int orientation = array.getInt(R.styleable.StockFieldView_or, 0);
            if (orientation == 0) {
                setOrientation(HORIZONTAL);
                fieldName.setLayoutParams(new LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 0.5f));
                fieldValue.setLayoutParams(new LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 0.5f));
                fieldValue.setGravity(Gravity.RIGHT);
            } else {
                setOrientation(VERTICAL);
                fieldName.setLayoutParams(new LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, 0, 0.5f));
                fieldValue.setLayoutParams(new LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, 0, 0.5f));
                fieldValue.setGravity(Gravity.LEFT);
            }
            setWeightSum(1f);
            final String name = getStringValue(context, array, R.styleable.StockFieldView_name);
            fieldName.setText(name);
            final float textSize = array.getDimensionPixelSize(R.styleable.StockFieldView_size, 20);
            fieldName.setTextSize(TypedValue.COMPLEX_UNIT_PX, textSize);
            fieldValue.setTextSize(TypedValue.COMPLEX_UNIT_PX, textSize * 0.8f);
            array.recycle();
        }
    }

    public StockFieldView(Context context, AttributeSet attrs, int defStyleAttr) {
        this(context, attrs);
    }

    public StockFieldView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        this(context, attrs);
    }

    public void setText(CharSequence text) {
        fieldValue.setText(text);
    }

    public void setTextColor(int color) {
        fieldValue.setTextColor(color);
    }

    private String getStringValue(Context context, TypedArray array, int stylelable) {
        String name = array.getString(stylelable);
        if (name == null) {
            final int stringId = array.getResourceId(stylelable, -1);
            if (stringId > 0) {
                name = context.getString(stringId);
            } else {
                name = "";
            }
        }
        return name;
    }

}

File: app/src/main/kotlin/com/github/premnirmal/ticker/ui/StockFieldView.kt
package com.github.premnirmal.ticker.ui

import android.content.Context
import android.content.res.TypedArray
import android.util.AttributeSet
import android.util.TypedValue
import android.view.Gravity
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import com.github.premnirmal.tickerwidget.R
import com.robinhood.ticker.TickerUtils
import com.robinhood.ticker.TickerView

/**
 * Created by premnirmal on 2/27/16.
 */
class StockFieldView @JvmOverloads constructor(
  context: Context,
  attrs: AttributeSet? = null
) : LinearLayout(context, attrs) {

  companion object {
    const val ORIENTATION_HORIZONTAL = 0
    const val ORIENTATION_VERTICAL = 1

    const val GRAVITY_LEFT = 0
    const val GRAVITY_RIGHT = 1
    const val GRAVITY_CENTER = 2
  }

  private val fieldname: TextView
  private val fieldvalue: TickerView

  init {
    LayoutInflater.from(context)
        .inflate(R.layout.stock_field_view, this, true)
    fieldname = findViewById(R.id.fieldname)
    fieldvalue = findViewById(R.id.fieldvalue)
    fieldvalue.setCharacterLists(TickerUtils.provideNumberList())
    attrs?.let {
      val array = context.obtainStyledAttributes(it, R.styleable.StockFieldView)
      val orientation = array.getInt(R.styleable.StockFieldView_or, 0)
      if (orientation == ORIENTATION_HORIZONTAL) {
        setOrientation(HORIZONTAL)
        fieldname.layoutParams =
          LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 0.5f)
        fieldvalue.layoutParams =
          LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 0.5f)
        fieldvalue.gravity = Gravity.END
      } else {
        setOrientation(VERTICAL)
        fieldname.layoutParams = LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        fieldvalue.layoutParams = LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        fieldvalue.gravity = Gravity.START
      }
      weightSum = 1f
      val name = getStringValue(context, array, R.styleable.StockFieldView_name)
      fieldname.text = name
      val textSize = array.getDimensionPixelSize(R.styleable.StockFieldView_size, 20)
          .toFloat()
      fieldname.setTextSize(TypedValue.COMPLEX_UNIT_PX, textSize)
      fieldvalue.setTextSize(textSize * 0.9f)
      val centerText = array.getBoolean(R.styleable.StockFieldView_center_text, false)
      when {
        centerText -> {
          fieldname.gravity = Gravity.CENTER
          fieldvalue.gravity = Gravity.CENTER
        }
        orientation == ORIENTATION_HORIZONTAL -> {
          fieldname.gravity = Gravity.START
          fieldvalue.gravity = Gravity.END
        }
        orientation == ORIENTATION_VERTICAL -> {
          val textGravity = array.getInt(R.styleable.StockFieldView_text_gravity, 0)
          when (textGravity) {
            GRAVITY_LEFT -> {
              fieldname.gravity = Gravity.START
              fieldvalue.gravity = Gravity.START
            }
            GRAVITY_RIGHT -> {
              fieldname.gravity = Gravity.END
              fieldvalue.gravity = Gravity.END
            }
            GRAVITY_CENTER -> {
              fieldname.gravity = Gravity.CENTER
              fieldvalue.gravity = Gravity.CENTER
            }
          }
        }
      }
      array.recycle()
    }
  }

  constructor(
    context: Context,
    attrs: AttributeSet,
    defStyleAttr: Int
  ) : this(context, attrs)

  constructor(
    context: Context,
    attrs: AttributeSet,
    defStyleAttr: Int,
    defStyleRes: Int
  ) : this(
      context, attrs
  )

  fun setLabel(text: CharSequence?) {
    fieldname.text = text
  }

  fun setText(text: CharSequence?) {
    fieldvalue.text = text.toString()
  }

  fun setTextColor(color: Int) {
    fieldvalue.textColor = color
  }

  private fun getStringValue(
    context: Context,
    array: TypedArray,
    stylelable: Int
  ): String {
    var name = array.getString(stylelable)
    if (name == null) {
      val stringId = array.getResourceId(stylelable, -1)
      name = if (stringId > 0) {
        context.getString(stringId)
      } else {
        ""
      }
    }
    return name
  }

}